### ISSUE TITLE
With faster scheduling, something in flux-core chokes

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-core#3625]
I am testing the performance and scalability gain from https://github.com/flux-framework/flux-sched/pull/820 and https://github.com/flux-framework/flux-sched/pull/826 on Quartz. 

One thing that I'm noticing is: as I am scaling up, it appears the faster scheduling performance leads to a bottleneck in other places in flux-core. I used a python script to asynchronously submit K 1-core `sleep (10)` jobs where K is the number of cores in the allocation. 64 nodes (i.e., 64 x 36 = 2304 cores/jobs ) are the first scale where this oddity showed up. For example, throttling job submission by injecting a delay of 500ms for every 50th job shortened the makespan by 1.73x. (99.83 sec vs. 57.66 secs). My scaling was a powers of two, starting from 2 nodes. 

I am adding my PerfFlowAspect analyses below. 

## 1. At 64 nodes with no throttling
![64N-no-throttle](https://user-images.githubusercontent.com/862123/116652801-c8621300-a93a-11eb-82c7-3eba09615023.png)

## 2. At 64 nodes with throttling -- injecting a delay of 500ms for every 50th job submission
![64N-throttle-500ms-50jobs](https://user-images.githubusercontent.com/862123/116653261-a9b04c00-a93b-11eb-9449-062f098889ed.png)

**Comment by grondo** [comment]
It would be interesting to know if the issue still occurs when you enable "mock" job execution by setting

```
    jobspec.setattr("system.exec.test.run_duration", runtime)
```

where `runtime` is the amount of time you want the job to sleep.

Since the `job-exec` module has to launch all 2000+ flux-shells, from the rank 0 module, I wonder if the bottleneck is in that portion of remote execution. This has already been identified as a problem in the 4th bullet of #3346.

**Comment by garlick** [comment]
Which version of `flux-core` is under test?

**Comment by dongahn** [comment]
@garlick:

```
quartz764{dahn}26: /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-c0.27.0.pre-s0.16.0.pre/bin/flux --version
commands:    		0.26.0-13-ge44070914
libflux-core:		0.26.0-13-ge44070914
build-options:		+hwloc==1.11.0
```

**Comment by dongahn** [comment]
Great info @grondo. I will do some more testing and update this ticket.

**Comment by dongahn** [comment]
While I'm waiting on the results for 64 Quartz Nodes, it seems pretty clear the execution system is part of this performance issue at scale:

##### Normal execution
```
 2 Nodes: Time elapsed:  13.770796537399292 secs
 4 Nodes: Time elapsed:  13.903920650482178 secs
 8 Nodes: Time elapsed:  14.903079509735107 secs
16 Nodes: Time elapsed:  17.32903003692627 secs
32 Nodes: Time elapsed:  28.66315484046936 secs
```

##### Emulated execution
```
 2 Nodes: Time elapsed:  12.943424940109253
 4 Nodes: Time elapsed:  13.18623948097229
 8 Nodes: Time elapsed:  13.890060186386108
16 Nodes: Time elapsed:  15.712011814117432
32 Nodes: Time elapsed:  19.800498247146606
```

**Comment by grondo** [comment]
Well it is difficult to interpret those results since I can't see the workload you've submitted, but there is a lot that is disabled when you use emulated execution. For example with the standard throughput test I can get 150 job/s throughput with emulated execution on my system, but only 40 job/s running real `sleep 0` jobs.

Emulated execution just measures how quickly the job-manager, scheduler, and job-exec module can move jobs through their state transition cycle. No job shell is actually launched, there is no execution and output eventlogs that need to be created, etc.

**Comment by dongahn** [comment]
Yeah... Just as a reference, I submit batch jobs on quartz:

```bash
#!/usr/bin/bash
module use /usr/global/tools/flux/toss_3_x86_64_ib/modulefiles
module load flux/perfflow
FLUX_BUILD=/usr/global/tools/flux/toss_3_x86_64_ib/build/2021-04-28-c0.27.0.pre-s0.16.0.pre
FLUX=/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-c0.27.0.pre-s0.16.0.pre/bin/flux
INITPROG=/g/g0/dahn/workspace/PACT21_evaluation/SmokeTest/initprog/submit_sleep.py
NJOBS=$(((SLURM_NNODES - 1) * 36))
INITPROG=/g/g0/dahn/workspace/PACT21_evaluation/SmokeTest/initprog/submit_sleep.py
export FLUXION_QMANAGER_OPTIONS='queue-params=queue-depth=32'
export FLUXION_RESOURCE_OPTIONS="load-allowlist=node,core,gpu policy=first"
export FLUX_EXEC_PATH_PREPEND=${FLUX_BUILD}/flux-sched/t/scripts

srun -N ${SLURM_NNODES} --mpi=none --mpibind=off ${FLUX} start <<-EOT
    flux resource drain 0
    flux resource list
    ${INITPROG} --njobs=${NJOBS} --sleep-second=10 ${INITPROG_ADDITIONAL_OPTS}
    flux ion-resource stat
EOT
```
submit_sleep.py is at https://gist.github.com/dongahn/eeee5f88445d88f3401c7770a4672eed

**Comment by dongahn** [comment]
> Emulated execution just measures how quickly the job-manager, scheduler, and job-exec module can move jobs through their state transition cycle. No job shell is actually launched, there is no execution and output eventlogs that need to be created, etc.

Understood. I am just measuring, collecting and logging those data point because there are currently users who are using Flux at scale in high throughput fashion and I want to provide them with best parameters and also suggests areas for future enhancements. 

The real issue for me is a long delay at 64 nodes (and probably at higher scales) that emerges if I don't do throttling and I'm trying to find where that comes from. Because this is something that these users can hit in their next large scale campaign own ORNL.

**Comment by dongahn** [comment]
> Just be sure to reload the scheduler and the resource module using the noverify module option

Hmmm, this method currently doesn't seem to work with fluxion. Need to create a ticket at flux-sched.

```
flux start
flux kvs put resource.R='{"version": 1, "execution": {"R_lite": [{"rank": "0-64", "children": {"core": "0-35"}}], "starttime": 0.0, "expiration": 0.0, "nodelist": ["quartz[31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31,31]"]}}'
flux module reload resource noverify
flux module load sched-fluxion-resource load-allowlist=node,core,gpu policy=first
flux module load sched-fluxion-qmanager queue-params=queue-depth=32
flux resource list
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1       36        0 quartz31
 allocated      0        0        0
      down      0        0        0
```

**Comment by dongahn** [comment]
Ah... maybe it is just that this R doesn't have JGF section...

**Comment by dongahn** [comment]
Now I have data point at 64 Quartz nodes. W/ emulated execution, the makespan improves by 2x. But there seems to other subsystem also involved in choking. Please see the graph below. 

##### Normal execution
```
 2 Nodes: Time elapsed:  13.770796537399292 secs
 4 Nodes: Time elapsed:  13.903920650482178 secs
 8 Nodes: Time elapsed:  14.903079509735107 secs
16 Nodes: Time elapsed:  17.32903003692627 secs
32 Nodes: Time elapsed:  28.66315484046936 secs
64 Nodes: Time elapsed: 99.83459043502808
```

##### Emulated execution
```
 2 Nodes: Time elapsed:  12.943424940109253 secs
 4 Nodes: Time elapsed:  13.18623948097229 secs
 8 Nodes: Time elapsed:  13.890060186386108 secs
16 Nodes: Time elapsed:  15.712011814117432 secs
32 Nodes: Time elapsed:  19.800498247146606 secs
64 Nodes: Time elapsed: 47.86593675613403 secs
```

Note that the first free request doesn't come to the scheduler until nearly 40 secs. Each job should only run 10 secs so there might be some performance event somewhere which makes execution system starving.

![Screen Shot 2021-04-30 at 5 53 53 PM](https://user-images.githubusercontent.com/862123/116766127-b71e1280-a9dd-11eb-98c3-218ccd900694.png)

**Comment by dongahn** [comment]
@grondo: I think I translated my test workloads to `throughput.py` as you suggested and ran three configurations in an exclusively allocated 1 quartz node (but with 64 fake nodes): 1) fluxion 2) sched-simple 3) sched-simple mode=unlimited. It seems `unlimited' mode has something to do with the potential starvation. 

```bash
#! /bin/bash

FLUXION_OUT=fluxion.out
SIMPLE_OUT=simple.out
SIMPLE_UNLIMITED_OUT=simple.unlimited.out

rm -f $FLUXION_OUT
rm -f $SIMPLE_OUT
rm -f $SIMPLE_UNLIMITED_OUT

touch $FLUXION_OUT
touch $SIMPLE_OUT
touch $SIMPLE_UNLIMITED_OUT

for iter in `seq 1 1`
do
    rm -f R.json
    flux R encode -r '0-63' -c '0-35' -H 'mock[0-63]' | flux ion-R encode > R.json
    flux start <<-EOT
    ./R-feeder R.json
    flux module remove sched-fluxion-resource
    flux module reload resource noverify
    flux module load sched-fluxion-resource load-allowlist=node,core,gpu policy=first
    flux module load sched-fluxion-qmanager queue-params=queue-depth=32
    flux resource list
    ./throughput.py -n 2304 -t 10 >> ${FLUXION_OUT}
EOT
done

for iter in `seq 1 1`
do
    rm -f R.json
    flux R encode -r '0-63' -c '0-35' -H 'mock[0-63]' > R.json
    flux start <<-EOT
    ./R-feeder R.json
    flux module remove sched-fluxion-resource
    flux module reload resource noverify
    flux module load sched-simple
    flux resource list
    ./throughput.py -n 2304 -t 10 >> ${SIMPLE_OUT}
EOT
done

for iter in `seq 1 1`
do
    rm -f R.json
    flux R encode -r '0-63' -c '0-35' -H 'mock[0-63]' > R.json
    flux start <<-EOT
    ./R-feeder R.json
    flux module remove sched-fluxion-resource
    flux module reload resource noverify
    flux module load sched-simple mode=unlimited
    flux resource list
    ./throughput.py -n 2304 -t 10 >> ${SIMPLE_UNLIMITED_OUT}
EOT
done
```

```
quartz1532{dahn}38: cat fluxion.out
number of jobs: 2304
submit time:    2.726 s (845.3 job/s)
script runtime: 60.901s
job runtime:    60.042s
throughput:     38.4 job/s (script:  37.8 job/s)

quartz1532{dahn}39: cat simple.out
number of jobs: 2304
submit time:    2.612 s (882.1 job/s)
script runtime: 39.922s
job runtime:    39.516s
throughput:     58.3 job/s (script:  57.7 job/s)

quartz1532{dahn}40: cat simple.unlimited.out
number of jobs: 2304
submit time:    3.020 s (763.0 job/s)
script runtime: 65.473s
job runtime:    63.603s
throughput:     36.2 job/s (script:  35.2 job/s)
```

My guess is the natural throttling that comes with `mode=single` helps avoiding some starvation somewhere...

**Comment by dongahn** [comment]
BTW, the fake scale test is awesome! I will go ahead and run thing at much larger scale as well.

**Comment by dongahn** [comment]
This is now a duplicate of Issue #3630.
